// oer-utils/dist/lib/predictor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("./util");
class Predictor {
    constructor() {
        this.size = 0;
    }
    get length() {
        return this.size;
    }
    writeUInt(value, length) {
        this.size += length;
    }
    writeInt(value, length) {
        this.size += length;
    }
    writeVarUInt(_value) {
        if (!(0, util_1.isInteger)(_value)) {
            throw new Error('UInt must be an integer');
        }
        let lengthOfValue;
        if (typeof _value === 'number') {
            if (_value < 0) {
                throw new Error('UInt must be positive');
            }
            lengthOfValue = (0, util_1.getUIntBufferSize)(_value);
        }
        else {
            const value = (0, util_1.longFromValue)(_value, true);
            lengthOfValue = (0, util_1.getLongUIntBufferSize)(value);
        }
        this.skipVarOctetString(lengthOfValue);
    }
    writeVarInt(_value) {
        if (!(0, util_1.isInteger)(_value)) {
            throw new Error('UInt must be an integer');
        }
        let lengthOfValue;
        if (typeof _value === 'number') {
            lengthOfValue = (0, util_1.getIntBufferSize)(_value);
        }
        else {
            const value = (0, util_1.longFromValue)(_value, false);
            lengthOfValue = (0, util_1.getLongIntBufferSize)(value);
        }
        this.skipVarOctetString(lengthOfValue);
    }
    writeOctetString(buffer, length) {
        if (buffer.length !== length) {
            throw new Error('Incorrect length for octet string (actual: ' +
                buffer.length +
                ', expected: ' +
                length +
                ')');
        }
        this.skip(length);
    }
    writeVarOctetString(buffer) {
        this.skipVarOctetString(buffer.length);
    }
    createVarOctetString(length) {
        this.skipVarOctetString(length);
        return new Predictor();
    }
    write(bytes) {
        this.size += bytes.length;
    }
    skip(bytes) {
        this.size += bytes;
    }
    getSize() {
        return this.size;
    }
    static measureVarOctetString(length) {
        let total = 1;
        if (length > 127) {
            const lengthOfLength = (0, util_1.getUIntBufferSize)(length);
            total += lengthOfLength;
        }
        total += length;
        return total;
    }
    skipVarOctetString(length) {
        this.skip(Predictor.measureVarOctetString(length));
    }
}
;
[8, 16, 32, 64].forEach((bits) => {
    Predictor.prototype[`writeUInt${bits}`] = function (value) {
        return this.writeUInt(value, bits / 8);
    };
    Predictor.prototype[`writeInt${bits}`] = function (value) {
        return this.writeUInt(value, bits / 8);
    };
});
exports.default = Predictor;
