// oer-utils/dist/lib/reader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const underflow_error_1 = require("../errors/underflow-error");
const parse_error_1 = require("../errors/parse-error");
const Long = require("long");
const util_1 = require("./util");
class Reader {
    constructor(buffer) {
        this.buffer = buffer;
        this.cursor = 0;
        this.bookmarks = [];
    }
    static from(source) {
        if (Buffer.isBuffer(source)) {
            return new Reader(source);
        }
        else if (source instanceof Reader) {
            return new Reader(source.buffer.slice(source.cursor));
        }
        else {
            throw new Error('Reader must be given a Buffer');
        }
    }
    bookmark() {
        this.bookmarks.push(this.cursor);
    }
    restore() {
        if (!this.bookmarks.length) {
            throw new Error('Cannot restore bookmark when no bookmark set');
        }
        this.cursor = this.bookmarks.pop();
    }
    ensureAvailable(bytes) {
        if (this.buffer.length < this.cursor + bytes) {
            throw new underflow_error_1.UnderflowError('Tried to read ' +
                bytes +
                ' bytes, but only ' +
                (this.buffer.length - this.cursor) +
                ' bytes available');
        }
    }
    readUIntNumber(length) {
        if (length < 1) {
            throw new Error('UInt length must be greater than zero');
        }
        else if (util_1.MAX_SAFE_BYTES < length) {
            throw new Error('Value does not fit a JS number without sacrificing precision');
        }
        else {
            const value = this.buffer.readUIntBE(this.cursor, length);
            this.cursor += length;
            return value;
        }
    }
    readUIntLong(length) {
        const value = this.peekUIntLong(length);
        this.cursor += length;
        return value;
    }
    readUInt(length) {
        if (length >= 1 && length <= util_1.MAX_SAFE_BYTES) {
            return String(this.readUIntNumber(length));
        }
        else {
            return this.readUIntLong(length).toString();
        }
    }
    peekUIntNumber(length) {
        if (length >= 1 && length <= util_1.MAX_SAFE_BYTES) {
            return this.buffer.readUIntBE(this.cursor, length);
        }
        else {
            throw new Error('Value does not fit a JS number without sacrificing precision');
        }
    }
    peekUIntLong(length) {
        if (length === 0) {
            return Long.UZERO;
        }
        else if (length < 0) {
            throw new Error('Tried to read integer with negative length (provided: ' + length + ')');
        }
        else if (length > 8) {
            throw new Error('UInts longer than 8 bytes must be encoded as VarUInts');
        }
        return util_1.bufferToLong(this.peek(length), true);
    }
    peekUInt(length) {
        if (length >= 1 && length <= util_1.MAX_SAFE_BYTES) {
            return String(this.peekUIntNumber(length));
        }
        else {
            return this.peekUIntLong(length).toString();
        }
    }
    skipUInt(length) {
        this.skip(length);
    }
    readIntNumber(length) {
        if (length < 1) {
            throw new Error('Int length must be greater than zero');
        }
        else if (util_1.MAX_SAFE_BYTES < length) {
            throw new Error('Value does not fit a JS number without sacrificing precision');
        }
        else {
            const value = this.buffer.readIntBE(this.cursor, length);
            this.cursor += length;
            return value;
        }
    }
    readIntLong(length) {
        const value = this.peekIntLong(length);
        this.cursor += length;
        return value;
    }
    readInt(length) {
        if (length >= 1 && length <= util_1.MAX_SAFE_BYTES) {
            return String(this.readIntNumber(length));
        }
        else {
            return this.readIntLong(length).toString();
        }
    }
    peekIntNumber(length) {
        if (length >= 1 && length <= util_1.MAX_SAFE_BYTES) {
            return this.buffer.readIntBE(this.cursor, length);
        }
        else {
            throw new Error('Value does not fit a JS number without sacrificing precision');
        }
    }
    peekIntLong(length) {
        if (length === 0) {
            return Long.ZERO;
        }
        else if (length < 0) {
            throw new Error('Tried to read integer with negative length (provided: ' + length + ')');
        }
        else if (length > 8) {
            throw new Error('Ints longer than 8 bytes must be encoded as VarInts');
        }
        return util_1.bufferToLong(this.peek(length), false);
    }
    peekInt(length) {
        if (length >= 1 && length <= util_1.MAX_SAFE_BYTES) {
            return String(this.peekIntNumber(length));
        }
        else {
            return this.peekIntLong(length).toString();
        }
    }
    skipInt(length) {
        this.skip(length);
    }
    readVarUIntNumber() {
        if (this.buffer[this.cursor] === 0) {
            throw new parse_error_1.ParseError('UInt of length 0 is invalid');
        }
        if (this.buffer[this.cursor] <= util_1.MAX_SAFE_BYTES) {
            return this.readUIntNumber(this.buffer[this.cursor++]);
        }
        else {
            throw new Error('Value does not fit a JS number without sacrificing precision');
        }
    }
    readVarUIntLong() {
        const buffer = this.readVarOctetString();
        if (buffer.length === 0) {
            throw new parse_error_1.ParseError('UInt of length 0 is invalid');
        }
        return util_1.bufferToLong(buffer, true);
    }
    readVarUInt() {
        return this.readVarUIntLong().toString();
    }
    peekVarUIntNumber() {
        this.bookmark();
        const value = this.readVarUIntNumber();
        this.restore();
        return value;
    }
    peekVarUIntLong() {
        this.bookmark();
        const value = this.readVarUIntLong();
        this.restore();
        return value;
    }
    peekVarUInt() {
        return this.peekVarUIntLong().toString();
    }
    skipVarUInt() {
        this.skipVarOctetString();
    }
    readVarIntNumber() {
        if (this.buffer[this.cursor] === 0) {
            throw new parse_error_1.ParseError('Int of length 0 is invalid');
        }
        if (this.buffer[this.cursor] <= util_1.MAX_SAFE_BYTES) {
            return this.readIntNumber(this.buffer[this.cursor++]);
        }
        else {
            throw new Error('Value does not fit a JS number without sacrificing precision');
        }
    }
    readVarIntLong() {
        const buffer = this.readVarOctetString();
        if (buffer.length === 0) {
            throw new parse_error_1.ParseError('Int of length 0 is invalid');
        }
        return util_1.bufferToLong(buffer, false);
    }
    readVarInt() {
        return this.readVarIntLong().toString();
    }
    peekVarIntNumber() {
        this.bookmark();
        const value = this.readVarIntNumber();
        this.restore();
        return value;
    }
    peekVarIntLong() {
        this.bookmark();
        const value = this.readVarIntLong();
        this.restore();
        return value;
    }
    peekVarInt() {
        return this.peekVarIntLong().toString();
    }
    skipVarInt() {
        this.skipVarOctetString();
    }
    readOctetString(length) {
        return this.read(length);
    }
    peekOctetString(length) {
        return this.peek(length);
    }
    skipOctetString(length) {
        return this.skip(length);
    }
    readLengthPrefix() {
        const length = this.readUInt8Number();
        if (length & Reader.HIGH_BIT) {
            const lengthPrefixLength = length & Reader.LOWER_SEVEN_BITS;
            const actualLength = lengthPrefixLength && this.readUIntNumber(lengthPrefixLength);
            const minLength = Math.max(128, 1 << ((lengthPrefixLength - 1) * 8));
            if (lengthPrefixLength === 0 || actualLength < minLength) {
                throw new parse_error_1.ParseError('Length prefix encoding is not canonical: ' +
                    actualLength +
                    ' encoded in ' +
                    lengthPrefixLength +
                    ' bytes');
            }
            return actualLength;
        }
        return length;
    }
    readVarOctetString() {
        const length = this.readLengthPrefix();
        return this.read(length);
    }
    peekVarOctetString() {
        this.bookmark();
        const value = this.readVarOctetString();
        this.restore();
        return value;
    }
    skipVarOctetString() {
        const length = this.readLengthPrefix();
        this.skip(length);
    }
    read(bytes) {
        this.ensureAvailable(bytes);
        const value = this.buffer.slice(this.cursor, this.cursor + bytes);
        this.cursor += bytes;
        return value;
    }
    peek(bytes) {
        this.ensureAvailable(bytes);
        return this.buffer.slice(this.cursor, this.cursor + bytes);
    }
    skip(bytes) {
        this.ensureAvailable(bytes);
        this.cursor += bytes;
    }
}
Reader.HIGH_BIT = 0x80;
Reader.LOWER_SEVEN_BITS = 0x7f;
;
['read', 'peek', 'skip'].forEach((verb) => {
    ;
    [1, 2, 4, 8].forEach((bytes) => {
        Reader.prototype[verb + 'UInt' + bytes * 8] = function () {
            return this[verb + 'UInt'](bytes);
        };
        Reader.prototype[verb + 'Int' + bytes * 8] = function () {
            return this[verb + 'Int'](bytes);
        };
        if (verb !== 'skip') {
            Reader.prototype[verb + 'UInt' + bytes * 8 + 'Number'] = function () {
                return this[verb + 'UIntNumber'](bytes);
            };
            Reader.prototype[verb + 'Int' + bytes * 8 + 'Number'] = function () {
                return this[verb + 'IntNumber'](bytes);
            };
            Reader.prototype[verb + 'UInt' + bytes * 8 + 'Long'] = function () {
                return this[verb + 'UIntLong'](bytes);
            };
            Reader.prototype[verb + 'Int' + bytes * 8 + 'Long'] = function () {
                return this[verb + 'IntLong'](bytes);
            };
        }
    });
});
exports.default = Reader;
