// ilp-logger/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatters = exports.Logger = void 0;
const debug_1 = __importDefault(require("debug"));
class Logger {
    constructor(namespace) {
        this.namespace = namespace;
        this.info = (0, debug_1.default)(namespace + ':info');
        this.warn = (0, debug_1.default)(namespace + ':warn');
        this.error = (0, debug_1.default)(namespace + ':error');
        this.debug = (0, debug_1.default)(namespace + ':debug');
        this.trace = (0, debug_1.default)(namespace + ':trace');
    }
    extend(namespace) {
        return new Logger(`${this.namespace}:${namespace}`);
    }
}
exports.Logger = Logger;
exports.formatters = debug_1.default.formatters;
const createLogger = function (namespace) {
    return new Logger(namespace);
};
createLogger.default = createLogger;
createLogger.Logger = Logger;
createLogger.formatters = debug_1.default.formatters;
exports.default = createLogger;
module.exports = createLogger;
